File: server/modules/auth/session.test.ts
import { describe, it, expect } from "vitest";
import { createSession, activateSession, deactivateSession, validateSession } from "./session.js";

describe("createSession", () => {
  it("returns a 64-char hex sid, inactive by default", () => {
    const session = createSession();
    expect(session.sid).toMatch(/^[0-9a-f]{64}$/);
    expect(session.active).toBe(false);
  });

  it("returns unique sids", () => {
    expect(createSession().sid).not.toBe(createSession().sid);
  });
});

describe("activateSession / deactivateSession", () => {
  it("activateSession sets active to true", () => {
    const session = createSession();
    activateSession(session);
    expect(session.active).toBe(true);
  });

  it("deactivateSession sets active to false", () => {
    const session = createSession();
    activateSession(session);
    deactivateSession(session);
    expect(session.active).toBe(false);
  });
});

describe("validateSession", () => {
  it("returns true for matching active sid", () => {
    const session = createSession();
    activateSession(session);
    expect(validateSession(session, session.sid)).toBe(true);
  });

  it("returns false for wrong sid", () => {
    const session = createSession();
    activateSession(session);
    expect(validateSession(session, "wrong")).toBe(false);
  });

  it("returns false when session is inactive", () => {
    const session = createSession();
    activateSession(session);
    deactivateSession(session);
    expect(validateSession(session, session.sid)).toBe(false);
  });
});

File: server-go/internal/auth/session_test.go
package auth

import (
	"regexp"
	"testing"
)

func TestCreateSession_Format(t *testing.T) {
	s := CreateSession()
	if !regexp.MustCompile(`^[0-9a-f]{64}$`).MatchString(s.SID()) {
		t.Errorf("expected 64-char hex SID, got %q", s.SID())
	}
	if s.Active() {
		t.Error("expected session to be inactive on creation")
	}
}

func TestCreateSession_Unique(t *testing.T) {
	if CreateSession().SID() == CreateSession().SID() {
		t.Error("expected unique sids")
	}
}

func TestActivate(t *testing.T) {
	s := CreateSession()
	s.Activate()
	if !s.Active() {
		t.Error("expected active after Activate()")
	}
}

func TestDeactivate(t *testing.T) {
	s := CreateSession()
	s.Activate()
	s.Deactivate()
	if s.Active() {
		t.Error("expected inactive after Deactivate()")
	}
}

func TestValidate_MatchingActiveSID(t *testing.T) {
	s := CreateSession()
	s.Activate()
	if !s.Validate(s.SID()) {
		t.Error("expected Validate to return true for matching active sid")
	}
}

func TestValidate_WrongSID(t *testing.T) {
	s := CreateSession()
	s.Activate()
	if s.Validate("wrong") {
		t.Error("expected Validate to return false for wrong sid")
	}
}

func TestValidate_Inactive(t *testing.T) {
	s := CreateSession()
	s.Activate()
	s.Deactivate()
	if s.Validate(s.SID()) {
		t.Error("expected Validate to return false when inactive")
	}
}
